src/libfsm/exec.c

```diff
@@ -18,7 +18,7 @@
 #include "internal.h"
 
 static struct fsm_state *
-nextstate(const struct fsm_state *state, char c)
+nextstate(const struct fsm_state *state, unsigned char c)
```

Comment:
`enum fsm_edge_type` in `src/libfsm/internal.h` defines `FSM_EDGE_EPSILON` as `UCHAR_MAX + 1`, so that won't fit in a char either way.

The sign extension bug seems to be happening truncating from an int to a char, and then expanding to the enum. If `nextstate`'s `c` argument is instead an `enum fsm_edge_type`, then the conversion from an int is correct.